R/analyse_pseudo_bulk.R

```diff
@@ -0,0 +1,373 @@
+#' @param group_by      entry in metadata table, based on these cluster annotation pseudo bulk is performed
+#' @param split_by      variable -> split by a variable within the metadata; k must be a string
+#'                      random -> splits based on a random number; k must be a number
+#'                      Louvain, Louvain_multilevel, SLM, Leiden -> subclusters k must be a list with [resolution, cluster_1, cluster_2]
+#' @param k_variable    variable dependent on the split_by
+#'
+#' @examples
+#' 
+#' #using SCE object
+#' SCE_OBJECT <- ZeiselBrainData()
+#' SCE_RESULT_RANDOM <- generate_pseudo_bulk_data(SCE_OBJECT, "level1class", "random",5)
+#' SCE_RESULT_VARIABLE <- generate_pseudo_bulk_data(SCE_OBJECT, "level1class","variable","tissue")
+#'
+#'
+#' @returns             returns pseudo bulk generated data
+#'
+#' @export
+setGeneric("generate_pseudo_bulk_data", function(object,
+                                                 group_by = NULL, 
+                                                 split_by  = "random", 
+                                                 k_variable = "4") standardGeneric("generate_pseudo_bulk_data"))
+
+#' @inherit generate_pseudo_bulk_data using Seurat
+#' @param group_by      entry in metadata table, based on these cluster annotation pseudo bulk is performed
+#' @param split_by      variable -> split by a variable within the metadata; k must be a string
+#'                      random -> splits based on a random number; k must be a number
+#'                      Louvain, Louvain_multilevel, SLM, Leiden -> subclusters k must be a list with [resolution, cluster_1, cluster_2]
+#' @param k_variable    variable dependent on the split_by
+#'
+#' @returns             returns pseudo bulk generated data
+setMethod("generate_pseudo_bulk_data", c("object" = "Seurat"), function(object, 
+                                                                        group_by, 
+                                                                        split_by = "random", 
+                                                                        k_variable){
+      
+  if (!is.character(split_by)) {
+    stop('Error: split_by must be a string e.g "variable", "random", "Louvain",...')
+  }
+    
+  if (split_by == "variable"){
+    if (!is.character(k_variable)) {
+      stop('Error: k must be a string')
+    }
+    result <- split_variable(seurat_object, group_by, k_variable)
+    return(result)
+  }
+  
+  if (split_by == "random"){
+    if (!is.numeric(k_variable)) {
+      stop('Error: k must be a number')
+    }
+    result <- split_variable_random(seurat_object, group_by, k_variable)
+    return(result)
+  }
+  
+  if(split_by == "Louvain"){
+    resolution_ <- k_variable[[1]]
+    subcluster_ref <- k_variable[[2]]   # subcluster variable depending on entry in metadata
+    subcluster_comp <- k_variable[[3]]  # subcluster variable depending on entry in metadata
+
+    if(!is.numeric(resolution_)){
```

Comment:
This check is repeated for all three algorithms. You could create one single check that is used once any of the three is selected (together with the check that `k_variable` is a list etc